Mathlib/NumberTheory/IntervalSums.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2025 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+
+import Mathlib.Algebra.Order.Ring.Star
+import Mathlib.Algebra.Order.Archimedean.Basic
+import Mathlib.Algebra.CharP.Defs
+import Mathlib.Topology.Algebra.Monoid.Defs
+import Mathlib.Topology.Algebra.InfiniteSum.Defs
+import Mathlib.Data.Int.Interval
+import Mathlib.Order.Filter.AtTopBot.Finset
+import Mathlib.Order.Filter.AtTopBot.Group
+import Mathlib.Algebra.Group.EvenFunction
+
+/-!
+# Sums over symmetric integer intervals
+
+This file contains some lemmas about sums over symmetric integer intervals `Ixx -N N` used in the
+definition of the Eisenstein series `E2`. In particular we define `SymmetricConditional`,
+`IcoFilter`, `IocFilter` as `SummationFilter`s  corresponding to the intervals
+`Icc -N N`, `Ico -N N`, `Ioc -N N` respectively.
+We also prove that these filters are all `NeBot` and `LeAtTop`.
+
+-/
+
+open Finset Topology Function
+
+@[to_additive]
+lemma Finset.prod_Icc_of_even_eq_range {α : Type*} [CommGroup α] {f : ℤ → α} (hf : f.Even) (N : ℕ) :
```

Comment:
These lemmas can probably be moved upstream somewhere? They're not specific to topological sums.